isobus/src/isobus_task_controller_client_objects.cpp

```diff
@@ -0,0 +1,472 @@
+		                           std::vector<std::uint8_t> deviceExtendedStructureLabel,
+		                           std::uint64_t clientIsoNAME) :
+		  Object(deviceDesignator, 0),
+		  serialNumber(deviceSerialNumber),
+		  softwareVersion(deviceSoftwareVersion),
+		  structureLabel(deviceStructureLabel),
+		  localizationLabel(deviceLocalizationLabel),
+		  extendedStructureLabel(deviceExtendedStructureLabel),
+		  NAME(clientIsoNAME)
+		{
+		}
+
+		std::string DeviceObject::get_table_id() const
+		{
+			return tableID;
+		}
+
+		ObjectTypes DeviceObject::get_object_type() const
+		{
+			return ObjectTypes::Device;
+		}
+
+		std::vector<std::uint8_t> DeviceObject::get_binary_object() const
+		{
+			std::vector<std::uint8_t> retVal;
+
+			retVal.reserve(31 +
+			               designator.size() +
+			               softwareVersion.size() +
+			               serialNumber.size() +
+			               extendedStructureLabel.size());
+
+			retVal.push_back(tableID[0]);
+			retVal.push_back(tableID[1]);
+			retVal.push_back(tableID[2]);
+			retVal.push_back(static_cast<std::uint8_t>(get_object_id() & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((get_object_id() >> 8) & 0xFF));
+			retVal.push_back(designator.size());
+			for (std::size_t i = 0; i < designator.size(); i++)
+			{
+				retVal.push_back(designator[i]);
+			}
+			retVal.push_back(softwareVersion.size());
+			for (std::size_t i = 0; i < softwareVersion.size(); i++)
+			{
+				retVal.push_back(softwareVersion[i]);
+			}
+			retVal.push_back(static_cast<std::uint8_t>(NAME & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 8) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 16) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 24) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 32) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 40) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 48) & 0xFF));
+			retVal.push_back(static_cast<std::uint8_t>((NAME >> 56) & 0xFF));
+			retVal.push_back(serialNumber.size());
+			for (std::size_t i = 0; i < serialNumber.size(); i++)
+			{
+				retVal.push_back(serialNumber[i]);
+			}
+			for (std::uint_fast8_t i = 0; i < MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH; i++)
+			{
+				if (i < structureLabel.size())
+				{
+					retVal.push_back(structureLabel[i]);
+				}
+				else
+				{
+					retVal.push_back(' ');
+				}
+			}
+			for (std::uint_fast8_t i = 0; i < MAX_STRUCTURE_AND_LOCALIZATION_LABEL_LENGTH; i++)
+			{
+				if (i < localizationLabel.size())
+				{
+					retVal.push_back(localizationLabel[i]);
+				}
+				else
+				{
+					retVal.push_back(' ');
+				}
+			}
+			//retVal.push_back(static_cast<std::uint8_t>(extendedStructureLabel.size()));
+			//for (std::size_t i = 0; i < extendedStructureLabel.size(); i++)
+			//{
+			//	retVal.push_back(extendedStructureLabel[i]);
+			//}
```

Comment:
Better to remove this commented out code I think?

Reply:
Oh, this reminds me I need to deal with the extended structure label for version4+, I can clean that up

Reply:
Added version 4 support in 3b09683f4c91639846d09444363652f1d044b037 which properly uses this code block